File: Java/Data Structures & Algorithms Module - Java/Week10/BSTAttempt2/BST.java
/*
BST.java
Billy Cussen
26/11/2020
*/
public class BST<T extends Comparable<T>> implements BSTInterface<T>{

    BSTNode<T> root;

    @Override
    public void insert(T element){
        if(isEmpty()){
            this.root = new BSTNode<T>(element);
        }
        insertHelper(element, this.root);
    }

    private void insertHelper(T element, BSTNode<T> current){
        if(element.compareTo(current.element)==-1){
            if(current.left==null){
                current.left = new BSTNode<T>(element);
            } else{
                insertHelper(element,current.left);
            }
        } else{
            if(current.right==null){
                current.right = new BSTNode<T>(element);
            } else{
                insertHelper(element,current.right);
            }
        }

    }
    
    @Override
    public void inOrder(){
        if(isEmpty()){
            throw new NullPointerException("Can't do in order traversal, BST is Empty!");
        }
        System.out.println("***BST IN ORDER TRAVERSAL***");
        inOrderHelper(this.root);
        System.out.println("***END***");
    }

    private void inOrderHelper(BSTNode<T> current){
        if(current==null){
            return;
        }
        inOrderHelper(current.left);
        System.out.println(current);
        inOrderHelper(current.right);
    }

    @Override
    public int size(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Size, BST is empty!");
        }
        return sizeHelper(this.root);
    }

    private int sizeHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.left)+sizeHelper(current.right);
    }

    @Override
    public int depth(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Depth, BST is empty!");
        }
        return depthHelper(this.root);
    }

    private int depthHelper(BSTNode<T> current){
        if(current==null){
            return 0;
        }

        int left = depthHelper(current.left);
        int right = depthHelper(current.right);

        if(left > right){
            return 1 + depthHelper(current.left);
        } else{
            return 1 + depthHelper(current.right);
        }
    }

    @Override
    public T getMin(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Min, BST is Empty!");
        }
        return minHelper(this.root);
    }

    private T minHelper(BSTNode<T> current){
        if(current.left==null){
            return current.element;
        }
        return minHelper(current.left);
    }

    @Override
    public T getMax(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Max, BST is Empty!");
        }
        return maxHelper(this.root);       
    }

    private T maxHelper(BSTNode<T> current){
        if(current.right==null){
            return current.element;
        }
        return maxHelper(current.right);
    }

    @Override
    public boolean isEmpty(){
        return this.root == null;
    }

    @Override
    public boolean contains(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't get Element, BST Is Empty!");
        }
        return containsHelper(this.root, element);
    }

    private boolean containsHelper(BSTNode<T> current, T element){
        if(current==null){
            return false;
        }

        if(element.compareTo(current.element)==0){
            return true;
        } else if(element.compareTo(current.element)==1){
            return containsHelper(current.right,element);
        } else{
            return containsHelper(current.left,element);
        }
    }

    @Override
    public T remove(T element){
        return null;
    }

    @Override
    public BSTNode<T> findNode(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't find Node, BST is Empty!");
        }
        return findNodeHelper(this.root, element);
    }

    private BSTNode<T> findNodeHelper(BSTNode<T> current, T element){
        if(current==null){
            return null;
        }

        if(element.compareTo(current.element)==0){
            return current;
        } else if(element.compareTo(current.element)==-1){
            return findNodeHelper(current.left,element);
        } else{
            return findNodeHelper(current.right,element);
        }
    }

    public BSTNode<T> findParent(T element){
        if(isEmpty()){
            throw new NullPointerException("Unable to find Parent, BST is Empty!");
        } else if(element.compareTo(this.root.element)==0){
            throw new NullPointerException("Can't get parent, Node selected is the Root!");
        }
        return findParentHelper(this.root, element);
    }

    private BSTNode<T> findParentHelper(BSTNode<T> current, T element){
        if(current==null){
            return null;
        }

        if(element.compareTo(current.element)==-1){
            if(element.compareTo(current.left.element)==0){
                return current;
            } else{
                return findParentHelper(current.left, element);
            }
        } else{
            if(element.compareTo(current.right.element)==0){
                return current;
            } else{
                return findParentHelper(current.right, element);
            }
        }
    }
}

File: Python/Data Structures & Algorithms Module - Python/Revision/BinarySearchTree/BST.py
"""
BST.py
Billy Cussen
18/01/2021
"""

from Node import Node

class BST:

    def __init__(self):
        self.root = None

    def size(self):
        if self.isEmpty():
            return 0
        return self.sizeHelper(self.root)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.left)+self.sizeHelper(current.right)

    def insert(self, element):
        if self.isEmpty():
            self.root = Node(element)
        else:
            self.insertHelper(element, self.root)

    def insertHelper(self, element, current):
        if ((element > current.element)-(element < current.element) == -1):
            if(current.left == None):
                current.left = Node(element)
            else:
                self.insertHelper(element, current.left)
        else:
            if(current.right == None):
                current.right = Node(element)
            else:
                self.insertHelper(element, current.right)

    def depth(self):
        if(self.isEmpty()):
            return 0
        else:
            return self.depthHelper(self.root)

    def depthHelper(self, current):
        if current == None:
            return 0

        leftNode = 1 + self.depthHelper(current.left)
        rightNode = 1 + self.depthHelper(current.right)
        if leftNode >= rightNode:
            return leftNode
        else:
            return rightNode

    def min(self):
        if self.isEmpty():
            raise Exception("Unable to get Min, BST is Empty!")
        else:
            return self.minHelper(self.root)

    def minHelper(self, current):
        if current == None:
            return None

        if current.left == None:
            return current.element
        else:
            return self.minHelper(current.left)

    def max(self):
        if self.isEmpty():
            raise Exception("Can't get Max, BST is Empty!")
        else:
            return self.maxHelper(self.root)

    def maxHelper(self,current):
        if current == None:
            return None

        if current.right == None:
            return current.element
        else:
            return self.maxHelper(current.right)

    def inOrder(self):
        if self.isEmpty():
            raise Exception("Unable to do in order traversal, BST is Empty!")
        print("***IN ORDER BST TRAVERSAL***\n")
        self.inOrderHelper(self.root)
        print("***END***\n")

    def inOrderHelper(self, current):
        if current == None:
            return
        
        self.inOrderHelper(current.left)
        print(current.toString())
        self.inOrderHelper(current.right)

    def isEmpty(self):
        return self.root == None